workflows/build.rs

```diff
@@ -0,0 +1,106 @@
+
+/// Generates Workflows as rust files from the yaml stored in /specs. Each Workflow is stored within
+/// its own mod within the `generated_workflows` module. Additionally, a function called `workflows`
+/// is generated that returns a vector of all the `Workflow`s that were created.
+fn main() -> Result<()> {
+    println!("cargo:rerun-if-changed=../specs");
+
+    std::fs::create_dir_all("src/generated_workflows")?;
+    let parent_module = std::fs::File::create("src/generated_workflows/mod.rs")?;
+
+    let mut workflows_added = Vec::new();
+
+    for entry in WalkDir::new("../specs").max_depth(1) {
+        let entry = entry?;
+
+        if entry.file_type().is_file() {
+            let file = File::open(entry.path())?;
+
+            let mmap = unsafe { memmap::Mmap::map(&file) }?;
+            let yaml_content = std::str::from_utf8(&mmap)?;
+
+            let workflow: Workflow = serde_yaml::from_str(yaml_content)?;
+            println!("generated workflow is {:?}", workflow);
+
+            let file_name = entry
+                .file_name()
+                .to_str()
+                .expect("OsStr should convert to str")
+                .replace(".yaml", "")
+                .to_case(Case::Snake);
+            println!("file name is {:?}", file_name);
+
+            // Create a module for each Workflow within the parent module.
+            workflows_added.push(file_name.clone());
+
+            write_workflow(workflow, file_name.as_str())?;
+
+            writeln!(&parent_module, "pub mod {};", file_name)?;
+        }
+    }
+
+    write_workflows_function(parent_module, &mut workflows_added)?;
+
+    println!("running cargo fmt...");
+    Command::new("cargo")
+        .args(["fmt", "-p", "warp-workflows"])
+        .output()?;
+
+    Ok(())
+}
+
+/// Writes a `workflows` function into the module at path `parent_modules`. The generated function
+/// will look approximately like:
+/// ```ignore
+/// use warp_workflows_metadata::*;
+/// pub fn workflows() -> Vec<Workflow> {
+///     vec![
+///         foo::workflow(),
+///         bar::workflow(),
+///         bazz:workflow(),
+///     ]
+/// }
+/// ```
+fn write_workflows_function(
+    mut parent_module: File,
+    workflows_added: &mut Vec<String>,
+) -> Result<()> {
+    writeln!(parent_module, "\nuse warp_workflows_types::*;")?;
+    writeln!(parent_module, "pub fn workflows() -> Vec<Workflow> {{")?;
+    writeln!(parent_module, "vec![")?;
+    for workflows in workflows_added {
+        writeln!(parent_module, "{}::workflow(),", workflows)?;
+    }
+    writeln!(parent_module, "]")?;
+    writeln!(parent_module, "}}")?;
+
+    Ok(())
+}
+
+/// Writes a Workflow into the path at `src/generated_workflows/<file_name>.rs`. The generated Rust
+/// will look like:
+/// ```ignore
+/// use warp_workflows_metadata::*;
+/// pub fn workflow -> Workflow {
+///    Workflow {
+///      ...
+///    }
+/// }
+/// ```
+fn write_workflow(workflow: Workflow, file_name: &str) -> Result<()> {
```

Comment:
would prefer to do this as a presubmit by validating the actual YAML instead of doing it at conversion time when converting from YAML -> rust, that we can we can ensure it's also correct before we bundle as an npm library

Reply:
yeah for sure, there are a few different presubmits we need (rust, npm, yaml). Going to add them as  followups